state-sync/storage-service/server/src/lib.rs

```diff
@@ -129,16 +269,80 @@ impl<T: StorageReaderInterface> StorageServiceServer<T> {
                 loop {
                     ticker.next().await;
 
+                    // Refresh the cache
                     if let Err(error) = refresh_cached_storage_summary(
-                        config,
-                        storage.clone(),
                         cached_storage_server_summary.clone(),
+                        storage.clone(),
+                        config,
                     ) {
                         let error = format!(
                             "Failed to refresh the cached storage summary! Error: {:?}",
                             error
                         );
-                        error!(LogSchema::new(LogEntry::StorageServiceError).message(&error));
+                        error!(LogSchema::new(LogEntry::StorageSummaryRefresh).message(&error));
+                    }
+                }
+            })
+            .await;
+    }
+
+    /// Spawns a non-terminating task that handles subscriptions
+    async fn spawn_subscription_handler(&mut self) {
+        let cached_storage_server_summary = self.cached_storage_server_summary.clone();
+        let config = self.config;
+        let data_subscriptions = self.data_subscriptions.clone();
+        let lru_storage_cache = self.lru_storage_cache.clone();
+        let storage = self.storage.clone();
+        let time_service = self.time_service.clone();
+
+        // Spawn the task
+        self.bounded_executor
+            .spawn(async move {
+                // Create a ticker for the refresh interval
+                let duration = Duration::from_millis(config.storage_summary_refresh_interval_ms);
+                let ticker = time_service.interval(duration);
+                futures::pin_mut!(ticker);
+
+                // Periodically check the data subscriptions
+                loop {
+                    ticker.next().await;
+
+                    // Remove all expired subscriptions
+                    remove_expired_data_subscriptions(config, data_subscriptions.clone());
+
+                    // Identify the peers with ready subscriptions
+                    let peers_with_ready_subscriptions = match get_peers_with_ready_subscriptions(
+                        cached_storage_server_summary.clone(),
```

Comment:
Can you clarify how often is this cache being updated? I am wondering if it makes sense for us to fetch the latest summary instead of using the cached one to avoid stale entry leading to higher response latency - since latency is the problem we are trying to solve with the subscription model.

Reply:
We don't need to fetch it per peer IIUC, you can just update the cache in the beginning of the loop, so that we can ensure that we have latest possible data before pushing to the client to reduce the overall e2e latency.

Reply:
Aah. That feels a little unclean to me. If we take your proposed model, then we end up in a world where we're doing 2 2 storage refreshes per `X` ms (the cache, and the subscription thread). Which begs the question, why not just update the cache twice as fast? i.e., halve `X`? 😄 That way, all the other things using the cache see the latest data.